src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -412,5 +467,107 @@ public void ChainHasAssumeValidHeaderAndMarkedForDownloadWhenPresented_SecondCha
             chainedHeaderDownloadTo = connectNewHeadersResult.DownloadTo;
             chainedHeaderDownloadTo.HashBlock.Should().Be(chainBTip.HashBlock);
         }
+
+        /// <summary>
+        /// Issue 16 @ Checkpoints are enabled. After the last checkpoint, there is an assumed valid. The chain
+        /// that covers them all is presented - marked for download. After that chain that covers the last checkpoint
+        /// but doesn't cover assume valid and is longer is presented - marked for download.
+        /// </summary>
+        [Fact]
+        public void ChainHasOneCheckPointAndAssumeValid_TwoAlternativeChainsArePresented_BothChainsAreMarkedForDownload()
+        {
+            // Chain header tree setup with disabled checkpoints.
+            // Initial chain has 2 blocks.
+            // Example: h1=h2.
+            const int initialChainSize = 2;
+            const int extensionChainSize = 2;
+            TestContext ctx = new TestContextBuilder().WithInitialChain(initialChainSize).UseCheckpoints().Build();
+            ChainedHeaderTree cht = ctx.ChainedHeaderTree;
+            ChainedHeader initialChainTip = ctx.InitialChainTip;
+
+            // Extend chain with 2 more headers.
+            initialChainTip = ctx.ExtendAChain(extensionChainSize, initialChainTip); // i.e. h1=h2=h3=h4
+            List<BlockHeader> listOfCurrentChainHeaders = ctx.ChainedHeaderToList(initialChainTip, initialChainSize + extensionChainSize);
+
+            // Setup a known checkpoint at header 4.
+            // Example: h1=h2=h3=(h4).
+            const int checkpointHeight = 4;
+            var checkpoint = new CheckpointFixture(checkpointHeight, listOfCurrentChainHeaders.Last());
+            ctx.SetupCheckpoints(checkpoint);
+
+            // Extend chain and add "Assume valid" at block 6.
+            // Example: h1=h2=h3=(h4)=h5=[h6].
+            const int chainExtension = 2;
+            ChainedHeader extendedChainTip = ctx.ExtendAChain(chainExtension, initialChainTip);
+            ctx.ConsensusSettings.BlockAssumedValid = extendedChainTip.HashBlock;
+
+            // Setup two alternative chains A and B. Chain A covers the last checkpoint (4) and "assume valid" (6).
+            // Chain B only covers the last checkpoint (4), but is longer that chain A.
```

Comment:
I would remove "but is longer" part from the comment because it is not relevant to what we want to test. It could be shorter or the same length and it would still be usable to test we want.

Reply:
That's what initial specs said. I can remove it if it is irrelevant. "Checkpoints are enabled. After last checkpoint there is assume valid. Chain that covers them all is presented- marked for download. After that chain that covers last checkpoint but doesnt cover assume valid and `is longer` is presented- marked for download." https://github.com/stratisproject/StratisBitcoinFullNode/issues/1321#issuecomment-389155904

Reply:
I think it is little bit of misunderstanding. I think what we wanted to say is that the alternative chain is longer than the height of assumed valid block, which means that is a block at that height and it is different and also there is some other block after that. The longer part was not meant to be in comparison to the first chain, but to the assumed valid height.

Reply:
Right, ok. Updated.